staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go

```diff
@@ -426,69 +427,101 @@ func (g *Cloud) AddAliasToInstance(nodeName types.NodeName, alias *net.IPNet) er
 // Gets the named instances, returning cloudprovider.InstanceNotFound if any
 // instance is not found
 func (g *Cloud) getInstancesByNames(names []string) ([]*gceInstance, error) {
+	instanceOrErrors, err := g.getInstanceOrErrorsByNames(names)
+	if err != nil {
+		return nil, err
+	}
+	var allInstances []*gceInstance
+	for _, entry := range instanceOrErrors {
+		if entry.err != nil {
+			return nil, entry.err
+		}
+		allInstances = append(allInstances, entry.instance)
+	}
+	return allInstances, nil
+}
+
+type instanceOrError struct {
+	instance *gceInstance
+	err      error
+}
+
+// Gets the named instances, returning a map of each name to either the found instances or
+// cloudprovider.InstanceNotFound if the instance is not found
+func (g *Cloud) getInstanceOrErrorsByNames(allNames []string) (map[string]*instanceOrError, error) {
 	ctx, cancel := cloud.ContextWithCallTimeout()
 	defer cancel()
 
-	found := map[string]*gceInstance{}
-	remaining := len(names)
-
-	nodeInstancePrefix := g.nodeInstancePrefix
-	for _, name := range names {
-		name = canonicalizeInstanceName(name)
-		if !strings.HasPrefix(name, g.nodeInstancePrefix) {
-			klog.Warningf("Instance %q does not conform to prefix %q, removing filter", name, g.nodeInstancePrefix)
-			nodeInstancePrefix = ""
-		}
-		found[name] = nil
-	}
+	found := map[string]*instanceOrError{}
+	allNamesSet := sets.NewString(allNames...)
+	remaining := len(allNames)
 
-	for _, zone := range g.managedZones {
-		if remaining == 0 {
-			break
+	for i := 0; i < len(allNames); i += 500 {
+		var names []string
+		if i+500 > len(allNames) {
+			names = allNames[i:]
+		} else {
+			names = allNames[i : i+500]
 		}
-		instances, err := g.c.Instances().List(ctx, zone, filter.Regexp("name", nodeInstancePrefix+".*"))
-		if err != nil {
-			return nil, err
+		var escapedInstanceNames []string
+		for _, name := range names {
+			canonicalizeName := canonicalizeInstanceName(name)
+			escapedInstanceNames = append(escapedInstanceNames, regexp.QuoteMeta(canonicalizeName))
+			found[canonicalizeName] = nil
 		}
-		for _, inst := range instances {
+		nameRegex := fmt.Sprintf("^(%s)$", strings.Join(escapedInstanceNames, "|"))
```

Comment:
This regexp is huge - do we know if it will work fine? Especially for projects with huge number of VMs from outside of the cluster?

I think this PR is mixing two separate things:
- allow to return partial result (the main purpose of this PR)
- switching to pagination (which is completely orthogonal problem)
I'm fairly confident that I don't want to do these two things in a single PR.

Can you please reuse the same logic that existed before (without pagination) and potentially open a separate PR to handle pagination - the latter has many more potential implications and I don't want to block this PR on those changes.

Reply:
From testing on the command line, the regex seems to work until it overflows the cmd lines buffer so it should work here. But just in case, i limit it to 500 instances at a time to reduce the size of the regex. The whole reason we need jt is that the scale tests were failing with the previous getinstancesbyname, which suggests that all instances are being listed. 

For a separate PR for pagination, do you want me to include the regex idea here or in the separate PR?

Reply:
> For a separate PR for pagination, do you want me to include the regex idea here or in the separate PR?

Yes - I just mean splitting into two PRs:
- one that adds new method, without really changing the logic of the existing function
- the second that does introduced this pagination

Reply:
I've updated this PR to only have the getInstanceOrErrorsByName change. The pagination and complicated regexp will be introduced in a separate PR after this. I still used a helper function to return instance or error because I don't want to change the contract of the existing getInstancesByName - tell me if you do.

Reply:
Yes - that is fine.
I added one comment - when applied lgtm.